environments/se/tests/src/test/java/org/jboss/weld/environment/se/test/event/enhanced/WeldEventTest.java

```diff
@@ -0,0 +1,172 @@
+            Assert.assertTrue(observer.isListObserved());
+
+            // select and fire from container.event()
+            observer.reset();
+            container.event().select(Bar.class, Dubious.Literal.INSTANCE).fire(new Bar());
+            Assert.assertTrue(observer.isBarObserved());
+
+            // and now with Types
+            Type someInterfaceType = SomeInterface.class;
+            Type listFooType = new TypeLiteral<List<Foo>>() {
+            }.getType();
+            Type listWildcardType = new TypeLiteral<List<?>>() {
+            }.getType();
+
+            // WeldEvent<Object> -> WeldEvent<SomeInterface> -> fire(SomeOtherBean)
+            observer.reset();
+            container.event().<SomeInterface>select(someInterfaceType).fire(container.select(SomeOtherBean.class).get());
+            Assert.assertTrue(observer.isSomeInterfaceObserved());
+            Assert.assertTrue(observer.isSomeOtherBeanObserved());
+            Assert.assertFalse(observer.isSomeTypedBeanObserved());
+
+            // WeldEvent<Object> -> WeldEvent<List<?>> -> fire(ArrayList<Object>)
+            observer.reset();
+            bean.getEventObject().select(listWildcardType).fire(new ArrayList<Object>());
+            Assert.assertFalse(observer.isListObjectObserved());
+            Assert.assertTrue(observer.isListObserved());
+
+            // WeldEvent<Object> -> WeldEvent<List<Foo>> -> fire(ArrayList<Foo>)
+            observer.reset();
+            bean.getEventObject().select(listFooType).fire(new ArrayList<Foo>());
+            Assert.assertFalse(observer.isListObjectObserved());
+            Assert.assertTrue(observer.isListObserved());
+        }
+    }
+
+    @Test
+    public void testEventSelectByType() {
+        try (WeldContainer container = new Weld().initialize()) {
+            BeanInjectingEvents bean = container.select(BeanInjectingEvents.class).get();
+
+            // prepare some Types
+            Type firstType = SomeInterface.class;
+            Type secondType = SomeOtherBean.class;
+            Type listFooType = new TypeLiteral<List<Foo>>() {
+            }.getType();
+            Type listWildcardType = new TypeLiteral<List<?>>() {
+            }.getType();
+
+            // verify invalid behaviour - selecting from Event<X> where X != Object
+            try {
+                // using Type from SomeInterface -> SomeOtherBean
+                bean.getEventSomeInterface().select(secondType);
+                Assert.fail();
+            } catch (IllegalStateException e) {
+                // expected
+            }
+
+            try {
+                // using Type from List<Object> -> List<?>
+                bean.getEventListObject().select(listWildcardType);
+                Assert.fail();
+            } catch (IllegalStateException e) {
+                // expected
+            }
+
+            // verify valid behaviour, following selections should work
+            bean.getEventObject().select(firstType);
```

Comment:
We should probably also test observability here... i.e. that a child `Event` works as expected?

Reply:
See `testEventObservability` method

Reply:
Ah, you're right ;-)